server/src/main/java/org/elasticsearch/search/lookup/SearchLookup.java

```diff
@@ -23,100 +23,62 @@
+        this.fieldTypeLookup = new TrackingMappedFieldsLookup(fieldTypeLookup);
         this.sourceLookup = new SourceLookup();
         this.fieldDataLookup = fieldDataLookup;
     }
 
-    /**
-     * Create a new {@link SearchLookup} that looks fields up the same as the one provided as argument,
-     * while also tracking field references starting from the provided field name. It detects cycles
-     * and prevents resolving fields that depend on more than {@link #MAX_FIELD_CHAIN_DEPTH} fields.
-     * @param searchLookup the existing lookup to create a new one from
-     * @param fieldChain the chain of fields that required the field currently being loaded
-     */
-    private SearchLookup(SearchLookup searchLookup, Set<String> fieldChain) {
-        this.fieldChain = Collections.unmodifiableSet(fieldChain);
+    private SearchLookup(SearchLookup searchLookup, String field) {
         this.sourceLookup = searchLookup.sourceLookup;
-        this.fieldTypeLookup = searchLookup.fieldTypeLookup;
+        this.fieldTypeLookup = searchLookup.fieldTypeLookup.excludingField(field);
         this.fieldDataLookup = searchLookup.fieldDataLookup;
     }
 
     /**
-     * Creates a copy of the current {@link SearchLookup} that looks fields up in the same way, but also tracks field references
-     * in order to detect cycles and prevent resolving fields that depend on more than {@link #MAX_FIELD_CHAIN_DEPTH} other fields.
-     * @param field the field being referred to, for which fielddata needs to be loaded
-     * @return the new lookup
-     * @throws IllegalArgumentException if a cycle is detected in the fields required to build doc values, or if the field
-     * being resolved depends on more than {@link #MAX_FIELD_CHAIN_DEPTH}
+     * Create a new {@link SearchLookup} that looks fields up the same as this one,
+     * while also tracking field references starting from the provided field name. It detects cycles
+     * and prevents resolving fields that depend on more than
+     * {@link TrackingMappedFieldsLookup#MAX_FIELD_CHAIN_DEPTH} fields.
+     * @param field        the field to exclude from further field lookups
      */
-    public final SearchLookup forkAndTrackFieldReferences(String field) {
-        Objects.requireNonNull(field, "field cannot be null");
-        Set<String> newFieldChain = new LinkedHashSet<>(fieldChain);
-        if (newFieldChain.add(field) == false) {
-            String message = String.join(" -> ", newFieldChain) + " -> " + field;
-            throw new IllegalArgumentException("Cyclic dependency detected while resolving runtime fields: " + message);
-        }
-        if (newFieldChain.size() > MAX_FIELD_CHAIN_DEPTH) {
-            throw new IllegalArgumentException("Field requires resolving too many dependent fields: " + String.join(" -> ", newFieldChain));
-        }
-        return new SearchLookup(this, newFieldChain);
+    public SearchLookup forkAndTrackReferences(String field) {
```

Comment:
I had hoped to be able to remove this method entirely, but we need it in runtime field queries if we want the root of the field cycle to be reported accurately.  In other words, without this method we still get errors thrown on cycles, but instead of being "field -> a -> b -> field" the cycle is reported as "a -> b -> field -> a".  I'm not sure if this is a real problem or not - it's a perfectly good cycle detection wherever it starts from.  I lean towards removing it anyway, as I think the additional API surface isn't worth the different error message.

Reply:
I kind of like the error message being accurate. Could we drop the API and have the queries make `TrackingMappedFieldsLookup` and immediate call the method to add themselves to its list?